### ISSUE TITLE
How to use Roswell to run code from my GitHub repo?

### ISSUE THREAD

**Issue by MatthewRock** [roswell/roswell#225]
Hi!

I am trying to use Roswell to install various CL implementations and to test my library that I have developed. However, Roswell's QL can't seem to find my library. I'm using Travis CI.

The library can be found [here](https://github.com/MatthewRock/cl-sandbox/).

Basically, here's what I do:

- Install Roswell using your script for CI.
- use `ros install MatthewRock/cl-sandbox` to install my library
- use `ros install fiveam` to install fiveam
- use my local ros script(essentially a main with quickload and asdf call to test my library)

I can't seem to figure out why, but this does not work.

Here's what I have tried:
- Linking my repository manually to the ~/.roswell/local-projects/ ; did not work.
- Linking my repository manually to the ~/.roswell/quicklisp/local-projects/ ; did not work.

Each time I'm getting QUICKLISP-CLIENT:SYSTEM-NOT-FOUND error.

What am I doing wrong?

**Comment by snmsts** [comment]
Thank you for your report.

What developers want to report is how we can reproduce your bug or phenomenon to fix.

- what kind of os or cpu or anything related your bug. 
- expression is also problem. what do you mean by "Install Roswell using your script for CI." We don't encourage users to use it for non-CI environment.
- Why don't you share your local ros script?
- Why no log in this report?

Those information are missing.
If you think we can notice what you expected by little information, that expectation is too much.
I feel sorry that I can say about it for now is only "I don't know".

**Comment by guicho271828** [comment]
Your ros scripts should be in the `roswell/` subdirectory of your repo.

**Comment by guicho271828** [comment]
ok, I see you have your scripts in `roswell` but it is in `travis-installation` branch.
`ros install MatthewRock/cl-sandbox/travis-installation` seems working fine.
@snmsts , however I noticed that the ros scripts are not recognized and they are not copied to `~/.roswell/bin`. I also noticed that if I run 
    ros install MatthewRock/cl-sandbox ; ros install MatthewRock/cl-sandbox/travis-installation
then the new branch is not checked out.

**Comment by guicho271828** [comment]
BTW I'm holding myself back from doing this fix; If I do, I would end up refactoring the entire code (which will eventually be reverted...).

**Comment by guicho271828** [comment]
BTW2 I understand the language issue but that sounds a bit harsh (you are a nice guy in Japanese I know). I sent a PR regarding this #226

**Comment by snmsts** [comment]
I admit.

**Comment by MatthewRock** [comment]
Thank you for your answers.

I hoped that providing links would be a nice way to not clutter topic at hand and provide you with enough details, but here it comes.

Example of two last builds, one when I try to get correct branch, one where not(all running ABCL installed by ros CI script):

Here is my Travis CI .yml script:

```
language: lisp
sudo: false

env:

  global:
    - PATH=~/.roswell/bin:$PATH
    - ROSWELL_INSTALL_DIR=$HOME/.roswell

  matrix:
    - LISP=abcl-bin
    - LISP=allegro
    - LISP=sbcl
    - LISP=cmu-bin
    - LISP=ccl-bin
    - LISP=clisp
    - LISP=ecl

addons:
  apt:
    packages:
      - libc6-i386
      - openjdk-7-jre

cache:
  directories:
    - $HOME/.roswell
    - $HOME/.config/common-lisp


install:
  - curl -L https://raw.githubusercontent.com/roswell/roswell/release/scripts/install-for-ci.sh | sh
  - ros install fiveam
  - ros install MatthewRock/cl-sandbox/travis-installation
script:
  - ros -s cl-sandbox -e "(asdf:test-system 'cl-sandbox)"

```

And here we've got roswell file:
```
#!/bin/sh
#|-*- mode:lisp -*-|#
#| Common Lisp sandbox package for testing out stuff.
exec ros -Q -- $0 "$@"
|#

(defun main (&rest argv)
  (ql:quickload '(:cl-sandbox) :silent t)
  (asdf:test-system 'cl-sandbox))
;;; vim: set ft=lisp lisp:
![screenshot-20170311 224114](https://cloud.githubusercontent.com/assets/1909226/23826770/2200fb62-06a4-11e7-8042-ef694d546e6b.png)
![screenshot-20170311 224243](https://cloud.githubusercontent.com/assets/1909226/23826771/2204b05e-06a4-11e7-8f19-aaf25bb2c446.png)

Screenshots are attached as files.

```

**Comment by guicho271828** [comment]
You are currently talking about two separate issues, (1) one about the scripts under `roswell` are not recognized when `ros install` from a git repo, (2) another about an ASDF error.

1. This is an actual issue in roswell, so should be fixed in the future.
2. Does the same problem happen for the other lisps? It seems so. I can reproduce the error on my local machine:

```sh
[guicho ~]$ ros install MatthewRock/cl-sandbox/travis-installation
install from github MatthewRock/cl-sandbox
building dump:roswell
up to date. stop
[guicho ~]$ ros -s cl-sandbox -e "(asdf:test-system 'cl-sandbox)"

debugger invoked on a QUICKLISP-CLIENT:SYSTEM-NOT-FOUND in thread
#<THREAD "main thread" RUNNING {1001BB72F3}>:
  System "cl-sandbox" not found

Type HELP for debugger help, or (SB-EXT:EXIT) to exit from SBCL.

restarts (invokable by number or by possibly-abbreviated name):
  0: [CONTINUE] Try again
  1: [ABORT   ] Give up on "cl-sandbox"
  2:            Ignore runtime option --eval "(ros:run '((:eval\"(ros:quicklisp)\")(:system \"cl-sandbox\")(:eval \"(asdf:test-system 'cl-sandbox)\")(:quit)))".
  3:            Skip rest of --eval and --load options.
  4:            Skip to toplevel READ/EVAL/PRINT loop.
  5: [EXIT    ] Exit SBCL (calling #'EXIT, killing the process).
```

Your asdf system definition file and the system name are **different**. You should rename `sandbox.asd` to `cl-sandbox.asd`.

**Comment by MatthewRock** [comment]
Thank you for your help and patience.

I would like to ask you whether you have tried the proposed solution. I must admit that asdf likes to work in unexpected ways, but on my personal PC with quicklisp properly set up, the file name was not an issue. I also changed the filename and pushed the change, but build still fails with the same error.

Instead, I would suggest that the problem is different one; when you install anything from github via Roswell, it uses filesystem to indicate where it came from(I believe that my installation would be named something like `~/.roswell/local-projects/MatthewRock/cl-sandbox/`). I am afraid that this might be problem, since I can't really see anything else that would be wrong with the setup.

Of course, I might be wrong. Do you have any ideas?

**Comment by guicho271828** [comment]
What I did is precisely the following:
```
$ cd ~/.roswell/local-projects
$ rm -rf MatthewRock
$ ros install MatthewRock/cl-sandbox
$ ros -s cl-sandbox -e "(asdf:test-system 'cl-sandbox)"    # fail
$ grep sandbox system-index.txt
MatthewRock/cl-sandbox/sandbox.asd
$ ros -s cl-sandbox -e "(asdf:test-system 'cl-sandbox)" # fail
$ grep sandbox system-index.txt
MatthewRock/cl-sandbox/sandbox.asd  # not updated
$ ros -e "(ql:register-local-projects)"
$ grep sandbox system-index.txt
MatthewRock/cl-sandbox/cl-sandbox.asd
$ ros -s cl-sandbox -e "(asdf:test-system 'cl-sandbox)"
; compiling file "/home/guicho/.roswell/local-projects/MatthewRock/cl-sandbox/t/sandbox-tests.lisp" 
.... Did 29 checks.
    Pass: 29 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)
```

I also tried the other branch. You still have a wrong file name :)

```
$ ls MatthewRock/cl-sandbox
ChangeLog  LICENSE  README.md  cl-sandbnox.asd  install.sh  roswell  src  t
```

**Comment by MatthewRock** [comment]
Oh snap. I overlooked the name... All works fine now. Thank you for your help.

**Comment by MatthewRock** [comment]
Hi. I don't want to open a new issue, since it kinda relates to the old one, so I thought I'd just reopen this one.

Right now I have noticed that the scheme I took previously - the one with the roswell - doesn't really work with (a) proper tests, (b) test package.

So let me explain (a); previously, I have been trying to test my system using `(asdf:test-system 'cl-sandbox)` . This doesn't really work; you always get T. However, this makes asdf look up how to test the system, realize that it has to get cl-sandbox/tests system, gets it and runs its test-op. Everything fine here.

So now, I have to change the way I do tests. Now I call the tests directly. HOWEVER, this does not work...

So the problem is: Roswell (or rather quicklisp) *doesn't see my tests system*! That's a huge problem because I can't invoke the tests. It's strange to me, since asdf can see the main package and test it.

Now what could be the cause? How should I set up the Travis's yml and roswell script so that the PRs are tested properly too?

For the reference, here are the links to 
- the repo: https://github.com/MatthewRock/cl-sandbox/tree/fix-tests/
- the travis file: https://github.com/MatthewRock/cl-sandbox/blob/fix-tests/.travis.yml
- said roswell file: https://github.com/MatthewRock/cl-sandbox/blob/fix-tests/roswell/cl-sandbox-test.ros

I have this repository locally on my machine under the path specified in `:tree` asdf's option. I can load both systems and run tests in my REPL without any problem.
